// @hyperlane-xyz/core/dist/factories/contracts/isms/routing/DomainRoutingIsmFactory.sol/AbstractDomainRoutingIsmFactory__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractDomainRoutingIsmFactory__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "contract DomainRoutingIsm",
                name: "module",
                type: "address",
            },
        ],
        name: "ModuleDeployed",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "uint32[]",
                name: "_domains",
                type: "uint32[]",
            },
            {
                internalType: "contract IInterchainSecurityModule[]",
                name: "_modules",
                type: "address[]",
            },
        ],
        name: "deploy",
        outputs: [
            {
                internalType: "contract DomainRoutingIsm",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class AbstractDomainRoutingIsmFactory__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.AbstractDomainRoutingIsmFactory__factory = AbstractDomainRoutingIsmFactory__factory;
AbstractDomainRoutingIsmFactory__factory.abi = _abi;
