// @hyperlane-xyz/sdk/dist/core/adapters/SealevelCoreAdapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SealevelCoreAdapter = void 0;
const web3_js_1 = require("@solana/web3.js");
const utils_1 = require("@hyperlane-xyz/utils");
const MultiProtocolApp_1 = require("../../app/MultiProtocolApp");
const ProviderType_1 = require("../../providers/ProviderType");
// https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/rust/sealevel/programs/mailbox/src/processor.rs
const MESSAGE_DISPATCH_LOG_REGEX = /Dispatched message to (.*), ID (.*)/;
class SealevelCoreAdapter extends MultiProtocolApp_1.BaseSealevelAdapter {
    constructor(chainName, multiProvider, addresses) {
        super(chainName, multiProvider, addresses);
        this.chainName = chainName;
        this.multiProvider = multiProvider;
        this.addresses = addresses;
    }
    extractMessageIds(sourceTx) {
        if (sourceTx.type !== ProviderType_1.ProviderType.SolanaWeb3) {
            throw new Error(`Unsupported provider type for SealevelCoreAdapter ${sourceTx.type}`);
        }
        const receipt = sourceTx.receipt;
        const logs = receipt.meta?.logMessages;
        if (!logs)
            throw new Error('Transaction logs required to check message delivery');
        const parsedLogs = SealevelCoreAdapter.parseMessageDispatchLogs(logs);
        if (!parsedLogs.length)
            throw new Error('Message dispatch log not found');
        return parsedLogs.map(({ destination, messageId }) => ({
            messageId: Buffer.from(messageId, 'hex').toString('hex'),
            destination: this.multiProvider.getChainName(destination),
        }));
    }
    async waitForMessageProcessed(messageId, destination, delayMs, maxAttempts) {
        const pda = SealevelCoreAdapter.deriveMailboxMessageProcessedPda(this.addresses.mailbox, messageId);
        const connection = this.multiProvider.getSolanaWeb3Provider(destination);
        await (0, utils_1.pollAsync)(async () => {
            // If the PDA exists, then the message has been processed
            // Checking existence by checking for account info
            const accountInfo = await connection.getAccountInfo(pda);
            if (accountInfo?.data?.length)
                return;
            else
                throw new Error(`Message ${messageId} not yet processed`);
        }, delayMs, maxAttempts);
    }
    static parseMessageDispatchLogs(logs) {
        const result = [];
        for (const log of logs) {
            if (!MESSAGE_DISPATCH_LOG_REGEX.test(log))
                continue;
            const [, destination, messageId] = MESSAGE_DISPATCH_LOG_REGEX.exec(log);
            if (destination && messageId)
                result.push({ destination, messageId });
        }
        return result;
    }
    /*
     * Methods for deriving PDA addresses
     * Should match https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/rust/sealevel/programs/mailbox/src/pda_seeds.rs
     */
    static deriveMailboxInboxPda(mailboxProgramId) {
        return super.derivePda(['hyperlane', '-', 'inbox'], mailboxProgramId);
    }
    static deriveMailboxOutboxPda(mailboxProgramId) {
        return super.derivePda(['hyperlane', '-', 'outbox'], mailboxProgramId);
    }
    static deriveMailboxDispatchedMessagePda(mailboxProgramId, uniqueMessageAccount) {
        return super.derivePda([
            'hyperlane',
            '-',
            'dispatched_message',
            '-',
            new web3_js_1.PublicKey(uniqueMessageAccount).toBuffer(),
        ], mailboxProgramId);
    }
    static deriveMailboxDispatchAuthorityPda(programId) {
        return super.derivePda(['hyperlane_dispatcher', '-', 'dispatch_authority'], programId);
    }
    static deriveMailboxProcessAuthorityPda(mailboxProgramId, recipient) {
        return super.derivePda([
            'hyperlane',
            '-',
            'process_authority',
            '-',
            new web3_js_1.PublicKey(recipient).toBuffer(),
        ], mailboxProgramId);
    }
    static deriveMailboxMessageProcessedPda(mailboxProgramId, messageId) {
        return super.derivePda(['hyperlane', '-', 'processed_message', Buffer.from(messageId, 'hex')], mailboxProgramId);
    }
}
exports.SealevelCoreAdapter = SealevelCoreAdapter;
