// @hyperlane-xyz/utils/dist/src/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InterchainSecurityModuleType = exports.MessageStatus = exports.ProtocolSmallestUnit = exports.ProtocolType = void 0;
var ProtocolType;
(function (ProtocolType) {
    ProtocolType["Ethereum"] = "ethereum";
    ProtocolType["Sealevel"] = "sealevel";
    ProtocolType["Fuel"] = "fuel";
})(ProtocolType || (exports.ProtocolType = ProtocolType = {}));
exports.ProtocolSmallestUnit = {
    [ProtocolType.Ethereum]: 'wei',
    [ProtocolType.Sealevel]: 'lamports',
};
var MessageStatus;
(function (MessageStatus) {
    MessageStatus[MessageStatus["NONE"] = 0] = "NONE";
    MessageStatus[MessageStatus["PROCESSED"] = 1] = "PROCESSED";
})(MessageStatus || (exports.MessageStatus = MessageStatus = {}));
var InterchainSecurityModuleType;
(function (InterchainSecurityModuleType) {
    InterchainSecurityModuleType[InterchainSecurityModuleType["MULTISIG"] = 3] = "MULTISIG";
})(InterchainSecurityModuleType || (exports.InterchainSecurityModuleType = InterchainSecurityModuleType = {}));
